internal/workload/wrapper.go

```diff
@@ -0,0 +1,130 @@
+package workload
+
+import (
+	"fmt"
+
+	"git.sr.ht/~spc/go-log"
+	api2 "github.com/jakub-dzon/k4e-device-worker/internal/workload/api"
+	"github.com/jakub-dzon/k4e-device-worker/internal/workload/network"
+	"github.com/jakub-dzon/k4e-device-worker/internal/workload/podman"
+	v1 "k8s.io/api/core/v1"
+)
+
+const nfTableName string = "edge"
+
+// workloadWrapper manages the workload and its configuration on the device
+type workloadWrapper struct {
+	workloads         *podman.Podman
+	netfilter         *network.Netfilter
+	mappingRepository *MappingRepository
+}
+
+func newWorkloadWrapper(configDir string) (*workloadWrapper, error) {
+	newPodman, err := podman.NewPodman()
+	if err != nil {
+		return nil, err
+	}
+	netfilter, err := network.NewNetfilter()
+	if err != nil {
+		return nil, err
+	}
+	mappingRepository, err := NewMappingRepository(configDir)
+	if err != nil {
+		return nil, err
+	}
+	return &workloadWrapper{
+		workloads:         newPodman,
+		netfilter:         netfilter,
+		mappingRepository: mappingRepository,
+	}, nil
+}
+
+func (ww workloadWrapper) Init() error {
+	return ww.netfilter.AddTable(nfTableName)
+}
+
+func (ww workloadWrapper) List() ([]api2.WorkloadInfo, error) {
+	infos, err := ww.workloads.List()
+	if err != nil {
+		return nil, err
+	}
+	for i := range infos {
+		infos[i].Name = ww.mappingRepository.GetName(infos[i].Id)
+	}
+	return infos, err
+}
+
+func (ww workloadWrapper) Remove(workloadName string) error {
+	if err := ww.workloads.Remove(ww.mappingRepository.GetId(workloadName)); err != nil {
+		return err
+	}
+	if err := ww.netfilter.DeleteChain(nfTableName, workloadName); err != nil {
+		log.Errorf("failed to delete chain %[1]s from %s table for workload %[1]s: %v", workloadName, nfTableName, err)
+	}
+	if err := ww.mappingRepository.Remove(workloadName); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (ww workloadWrapper) Run(workload *v1.Pod, manifestPath string) error {
+	if err := ww.applyNetworkConfiguration(workload); err != nil {
+		return err
+	}
+	podIds, err := ww.workloads.Run(manifestPath)
+	if err != nil {
+		return err
+	}
+	return ww.mappingRepository.Add(workload.Name, podIds[0])
```

Comment:
I wonder what would happen if workload was run successfully and we failed to add mapping. I think we need to make it transactional either both are successful or none.

Reply:
I agree with Moti - the complexity of compensating for errors in the `Run` method will only grow and the workload watchdog is a good place to make sure the state of the device is correct.

Reply:
I will change the `ensureWorkloadsFromManifestsAreRunning` method appropriately.

Reply:
Fixed. @masayag, @pkliczewski - please have a look.